src/algebra/lie_algebras.lean

```diff
@@ -0,0 +1,192 @@
+/-
+Copyright (c) 2019 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import ring_theory.algebra data.matrix.basic
+
+/-!
+# Lie algebras
+
+This file defines Lie rings, and Lie algebras over a commutative ring. It shows how these arise from
+associative rings and algebras via the ring commutator. In particular it defines the Lie algebra
+of endomorphisms of a module as well as of the algebra of square matrices over a commutative ring.
+
+## Notations
+
+We introduce the notation ⁅x, y⁆ for the Lie bracket. Note that these are the Unicode "square with
+quill" square brackets rather than the usual square brackets.
+
+## Implementation notes
+
+Lie algebras are defined as modules with a compatible Lie ring structure, and thus are partially
+unbundled. Since they extend Lie rings, these are also partially unbundled.
+
+## References
+* [N. Bourbaki, *Lie Groups and Lie Algebras, Chapters 1--3*][bourbaki1975]
+
+## Tags
+
+lie bracket, ring commutator, jacobi identity, lie ring, lie algebra
+-/
+
+universes u v
+
+class has_bracket (L : Type v) := (bracket : L → L → L)
+
+notation `⁅`x`,` y`⁆` := has_bracket.bracket x y
+
+namespace ring_commutator
+
+variables (A : Type v) [ring A]
+
+instance bracket : has_bracket A :=
+{ bracket := λ x y, x*y - y*x }
+
+@[simp] lemma add_left (x y z : A) :
+  ⁅x + y, z⁆ = ⁅x, z⁆ + ⁅y, z⁆ :=
+by { unfold has_bracket.bracket, simp [right_distrib, left_distrib] }
+
+@[simp] lemma add_right (x y z : A) :
+  ⁅z, x + y⁆ = ⁅z, x⁆ + ⁅z, y⁆ :=
+by { unfold has_bracket.bracket, simp [right_distrib, left_distrib] }
+
+@[simp] lemma alternate (x : A) :
+  ⁅x, x⁆ = 0 :=
+by { unfold has_bracket.bracket, simp }
+
+lemma jacobi (x y z : A) :
+  ⁅x, ⁅y, z⁆⁆ + ⁅y, ⁅z, x⁆⁆ + ⁅z, ⁅x, y⁆⁆ = 0 :=
+begin
+  unfold has_bracket.bracket,
+  repeat { rw mul_sub_left_distrib },
+  repeat { rw mul_sub_right_distrib },
+  repeat { rw add_sub },
+  repeat { rw ←sub_add },
+  repeat { rw ←mul_assoc },
+  have h : ∀ (x y z : A), x - y + z + y = x+z := by simp,
+  repeat { rw h },
+  simp,
+end
+
+end ring_commutator
+
+/--
+A Lie ring is an additive group with compatible product, known as the bracket, satisfying the
+Jacobi identity. The bracket is not associative unless it is identically zero.
+-/
+class lie_ring (L : Type v) [add_comm_group L] extends has_bracket L :=
+(add_left : ∀ (x y z : L), ⁅x + y, z⁆ = ⁅x, z⁆ + ⁅y, z⁆)
+(add_right : ∀ (x y z : L), ⁅z, x + y⁆ = ⁅z, x⁆ + ⁅z, y⁆)
+(alternate : ∀ (x : L), ⁅x, x⁆ = 0)
+(jacobi : ∀ (x y z : L), ⁅x, ⁅y, z⁆⁆ + ⁅y, ⁅z, x⁆⁆ + ⁅z, ⁅x, y⁆⁆ = 0)
+
+namespace lie_ring
+
+variables (L : Type v) [add_comm_group L] [lie_ring L]
+
+attribute [simp] add_left
+attribute [simp] add_right
+attribute [simp] alternate
+attribute [simp] jacobi
+
+@[simp] protected lemma skew (x y : L) :
```

Comment:
I've been thinking a bit about naming things. We have things like `add_zero` for `a + 0 = a`, and `mul_comm` for `a * b = b * a`. These things exist in the global namespace, which is really useful for having concise names. I imagine it would be useful to do something similar here. So here is my proposal. We need a {pre,suf}fix for talking about the lie bracket, similar to the `add_` and `mul_` and `smul_` prefixes, etc... (sometimes they appear as suffix, eg `one_mul`).

1. `bracket_` is too long,
2. `bra_` has (unfortunate) connotations
3. `lie_` seems pretty clear and to the point.

Therefore I suggest turning `namespace lie_ring` into `section lie_ring` on line 84.
And this lemma would be called `lie_skew`.